packages/gatsby-link/src/index.js

```diff
@@ -9,29 +9,44 @@ export { parsePath }
 
 const isAbsolutePath = path => path?.startsWith(`/`)
 
-export function withPrefix(path, prefix = __BASE_PATH__) {
+export function withPrefix(path, prefix = getGlobalBasePrefix()) {
   if (!isLocalLink(path)) {
     return path
   }
 
   if (path.startsWith(`./`) || path.startsWith(`../`)) {
     return path
   }
-  const base = prefix ?? __PATH_PREFIX__ ?? `/`
+  const base = prefix ?? getGlobalPathPrefix() ?? `/`
 
   return `${base?.endsWith(`/`) ? base.slice(0, -1) : base}${
     path.startsWith(`/`) ? path : `/${path}`
   }`
 }
 
+// These global values are wrapped in typeof clauses to ensure the values exist.
+// This is especially problematic in unit testing of this component.
+const getGlobalPathPrefix = () =>
+  process.env.NODE_ENV !== `production`
+    ? typeof __PATH_PREFIX__ !== `undefined`
+      ? __PATH_PREFIX__
+      : undefined
+    : __PATH_PREFIX__
```

Comment:
Nested ternaries are really hard to read. How about

```suggestion
const getGlobalPathPrefix = () => {
    if( process.env.NODE_ENV === `production` ) {
        return undefined
    }
    return __PATH_PREFIX__ ?? undefined
}
```

Reply:
Ironically, the ternary confused me so I misunderstood what was happening! I see what you're doing, but I'd still rather replafe the ternary with an early return.

Reply:
In the context of this discussion; I prefer to use `===` over `!==` if there is an `else` branch unless there's a clear reason to show the `!==` branch first. I think it reads better as inverting adds to the mental overhead of parsing code. I agree with Matt that this one is a little hairy. Sometimes it is what it is.

I would consider to ignore the production check (what purpose does that serve?) since it currently assumes the var exists in production anyways, so `typeof __PATH_PREFIX__ === 'undefined'` should be false regardless. So I'd suggest this instead.

```suggestion
const getGlobalPathPrefix = () =>
    // Prevent reference error; the `__PATH_PREFIX__` binding may not exist in tests
    typeof __PATH_PREFIX__ === `undefined` ? undefined : __PATH_PREFIX__
```